lib/seahorn/Houdini.cc

```diff
@@ -0,0 +1,292 @@
+			  get_all_bvars(body_cand, std::back_inserter(body_bvars));
+			  for(ExprVector::iterator it_bv = body_bvars.begin(); it_bv != body_bvars.end(); ++it_bv)
+			  {
+				  unsigned bvar_id = bind::bvarId(*it_bv);
+				  Expr body_app_arg = bind::domainTy(body_pred_app, bvar_id);// To improve
+				  body_bvar_map.insert(std::pair<Expr,Expr>(*it_bv, body_app_arg));
+			  }
+			  Expr body_cand_app = replace(body_cand, body_bvar_map);
+			  //outs() << "[BODY CAND APP]: " << *body_cand_app << "\n";
+			  body_map.insert(std::pair<Expr, Expr>(body_pred_app, body_cand_app));
+		  }
+		  Expr new_body = replace(body, body_map);
+
+		  //outs() << "[NEW BODY]: " << *new_body << "\n";
+
+		  //update cand and cand_app
+		  Expr cand_app = mk<AND>(neg_head_cand_app, new_body);
+
+		  outs() << "[CANDIDATE]: " << *cand_app << "\n";
+
+		  if(validateRule(cand_app, hm) == SAT)
+		  {
+			 //add rules in db.use(r.head()) to worklist
+			 Expr head_app = r.head();
+			 Expr head_fdecl = bind::fname(head_app);
+			 outs() << "[USE SET SIZE]: " << db.use(head_fdecl).size() << "\n";
+			 for(auto r_use : db.use(head_fdecl))
+		     {
+				 bool isExist = false;
+				 for(HornClauseDB::RuleVector::iterator itr=workList.begin(); itr!=workList.end(); ++itr)
+				 {
+					 if(*itr == *r_use)
+					 {
+						 isExist = true;
+						 break;
+					 }
+				 }
+				 if(isExist == false)
+				 {
+					 workList.push_back(*r_use);
+				 }
+			 }
+			 //weaken candidate for r.head()
+			 while (validateRule(cand_app, hm) == SAT)
+			 {
+				 //outs() << "[ARITY]: " << head_cand_app->arity() << "\n";
+				 if(isOpX<TRUE>(head_cand_app))
+			     {
+					 break;
+				 }
+				 if(!isOpX<AND>(head_cand_app))
+				 {
+					 head_cand_app = mk<TRUE>(head_cand_app->efac());
+				 }
+				 else
+				 {
+					 ExprVector new_head_vec;
+					 for(auto it = head_cand_app->args_begin (), end = head_cand_app->args_end (); it != end; ++it)
+					 {
+					 	new_head_vec.push_back(*it);
+					 }
+					 new_head_vec.erase(new_head_vec.begin());
+					 if(new_head_vec.size() > 1)
+					 {
+					 	head_cand_app = mknary<AND>(new_head_vec.begin(), new_head_vec.end());
+					 }
+					 else
+					 {
+					 	head_cand_app = new_head_vec[0];
+					 }
+				 }
+				 neg_head_cand_app = mk<NEG>(head_cand_app);
+			     //outs() << "[HEAD AFTER WEAKEN]: " << *neg_head_cand_app << "\n";
+				 cand_app = mk<AND>(neg_head_cand_app, new_body);
+				 outs() << "[CANDIDATE AFTER WEAKEN]: " << *cand_app << "\n";
+			  }
+			  workList.erase(workList.begin());
+		  }
+		  else // the ret is UNSAT
+		  {
+		      workList.erase(workList.begin());
+		  }
+	  }
+  }
+
+  bool Houdini::validateRule(Expr cand_app, HornifyModule &hm)
+  {
+	  // should call smt solver
+	  ZSolver<EZ3> solver(hm.getZContext ());
+	  solver.assertExpr(cand_app);
+	  solver.toSmtLib(outs());
+	  if(solver.solve())
+	  {
+		  outs() << "SAT\n";
+	  }
+	  else
+	  {
+		  outs() << "UNSAT\n";
+	  }
+	  return solver.solve();
```

Comment:
should not call `solve()`. Can be expensive and can change state.

Need a better architecture where the solver is not being destroyed and re-created each time.

I don't see where the transition relation (the body of the rule) is loaded. Perhaps `validateRule()` is a misleading name since the input is not a rule but an expression.